### ISSUE TITLE
ReadDirectory change proposal

### ISSUE THREAD

**Issue by billziss-gh** [billziss-gh/winfsp#34]
I have spent considerable time thinking about the `ReadDirectory` operation. `ReadDirectory` is currently defined as:

    NTSTATUS (*ReadDirectory)(FSP_FILE_SYSTEM *FileSystem,
        PVOID FileContext, PVOID Buffer, UINT64 Offset, ULONG Length,
        PWSTR Pattern,
        PULONG PBytesTransferred);

`ReadDirectory` was modelled after `fuse_lowlevel_ops::readdir`. In particular the `Offset` parameter allows a file system to return a directory listing in "chunks" ("pages"). The `Offset` parameter can be used to "seek" into a specific position into the directory stream and start reading from there. The kernel does not interpret the `Offset` parameter other than to specify that the value `0` means to start from the beginning of the directory stream. The kernel deduces the `Offset` to send to the user mode file system from the `FSP_FSCTL_DIR_INFO::NextOffset` fields.

Experience with implementing multiple file systems has shown that `ReadDirectory` is one of the harder operations to implement because of the `Offset` parameter. The problem is that many file systems do not have a concept of a numerical offset for directories. One solution seems to use a directory entry's file name as a marker/offset, which must be then somehow converted or mapped to a numerical offset to be returned to the kernel. Another solution is to buffer all directory contents and then use offsets within the buffer as the directory offsets; but this has other problems (among others, it has subtle consistency issues when directories are concurrently modified with multiple directory handles opened).

I have come to the realization that a much better definition for `ReadDirectory` would be:

    NTSTATUS (*ReadDirectory)(FSP_FILE_SYSTEM *FileSystem,
        PVOID FileContext, PVOID Buffer, ULONG Length,
        PWSTR Pattern,
        PWSTR Marker,
        PULONG PBytesTransferred);

Basically the `Marker` parameter would be `NULL` or the file name of the last directory entry seen by the kernel. The file system would have to return the first directory entry **after** the `Marker` or the first entry if `Marker` is `NULL`.

This change assumes that the file system always returns directory entries in the same order for a particular directory. This order does not have to be alphabetical or lexicographical, but it does have to be consistent. It should also be possible/easy for a file system to find all directory entries with file names greater than the `Marker` in the directory order.

I believe the `Marker` would simplify returning directory entries for many file systems.

For example, the MEMFS file system already maintains files in alphabetical order and it currently has to maintain a mapping of offsets to file names to implement `ReadDirectory`. The MEMFS `ReadDirectory` code would be greatly simplified if a `Marker` was passed instead.

As another example, in my (currently closed-source) file system SECFS, directory entries are maintained by the order of the cryptographic hash of their filenames. In order to find the directory entry that is greater than the `Marker` I have to hash it using the crypto-hash and then find all hashes that are greater than the `Marker` hash.

I am therefore proposing that `ReadDirectory` be changed to have a `Marker` instead of an `Offset`. I am currently experimenting with this change to see if it indeeds makes file system implementation easier. If anyone has strong opinions on this, feel free to voice them.

**Comment by billziss-gh** [comment]
I have started implementing this change in the pvt-ReadDirectory-change branch. I have also implemented `FspFileSystem*DirectoryBuffer` API's that make working with a `Marker` relatively simple for file systems that cannot maintain a consistent directory order on their own or that cannot seek by file name within a directory.

I have reimplemented the MEMFS and Passthrough file systems and this change seems to be an overall win to me. The code of both file systems is simpler and easier to understand IMO.

This is work in progress. The FUSE layer needs to be reimplemented to use `Marker`'s.
